### ISSUE TITLE
wireguard new tunnel configuration tutorial

### ISSUE THREAD

**Issue by highfreq** [Ysurac/openmptcprouter#2123]
In this tutorial i'll add a tunnel on the vps and configure a client to connect to it. I don't pretend it to be perfect but it works and i can ping the tunnel ends.

Server side

Copy configuration 

cp /etc/wireguard/wg0.conf /etc/wireguard/wg1.conf

Generate server keys on server machine

cd /etc/wireguard

wg genkey | tee vpn-server-privatekey-wg1.key | wg pubkey > vpn-server-public-wg1.key

Generate client keys on client machine with following command

wg genkey | tee vpn-client-privatekey-wg1.key | wg pubkey > vpn-client-public-wg1.key

Configure server

sudo nano /etc/wireguard/wg1.conf

Fill server private key with result of

cat /etc/wireguard/vpn-server-privatekey-wg1.key

Fill client private key with result of following command executefd on client machine

cat /etc/wireguard/vpn-client-public-wg1.key

Chose server address in my example i used 10.255.230.1/24
Chose client address in my example i used 10.255.230.2/24

[Interface]
Address = 10.255.230.1/24
ListenPort = 65312
PrivateKey = server private key

[Peer]
PublicKey = client public key
AllowedIPs = 10.255.230.0/24

Save and exit

Configure client

sudo nano /etc/wireguard/wg1.conf

Fill client private field with result of

cat /etc/wireguard/vpn-client-privatekey-wg1.key

Fill server public key with result of following command executefd on server machine

cat /etc/wireguard/vpn-client-public-wg1.key

Fill server public and port number with your server public ip and port number you configured on server side

[Interface]
Address = 10.255.230.2/24
PrivateKey = client private key
DNS = 1.1.1.1

[Peer]
PublicKey = server public key
Endpoint = server public ip : port number
AllowedIPs = 10.255.230.0/24

Save and exit

start new instances on server and client with following command

sudo systemctl start wg-quick@wg1.service

If you want to start your new instance at boot issue following command

sudo systemctl enable wg-quick@wg1.service

On both machines issuing ifconfig you should se the new wg1 tunnel

test connection by pinging 10.255.230.1 from client or 10.255.230.2 from server

This configuration will route all your 10.255.230.0 trafic trough the wg1 tunnel

Hope it is clear enough :)

**Comment by highfreq** [comment]
One thing i now need to have my configuration 100% working is some help on how to configure shorewall and enable my 10.255.230.2 client on wg to connect to to OMR ports.

**Comment by highfreq** [comment]
Any hint thatm i could take as an example to work on?

**Comment by Ysurac** [comment]
Maybe something like:
```
DNAT           vpn:10.255.230.1:2223             vpn:$OMR_ADDR   tcp     2223
```
But I'm not exactly sure about what you want ;)

**Comment by highfreq** [comment]
Done it exaclty the way you described it (i am using glorytin tcp) but pc in omr lan can't ping 10.255.230.2
I can ping 10.255.230.1 but not the other side of the tunnel.

**Comment by Ysurac** [comment]
You may need to change in /etc/shorewall/policy the line ```vpn     vpn   DROP``` to ```vpn    vpn ACCEPT```

**Comment by highfreq** [comment]
@Ysurac Any chance i can get some help in reaching the OMR network from the new wireguard tunnel? Would really need it and really appreciate any help.

**Comment by Ysurac** [comment]
This should work.
Try to use tcpdump on VPS and on router to check if you get something.
Also you can enable shorewall log to check if it's not blocked.

**Comment by highfreq** [comment]
how do i enable shorewall log?

**Comment by highfreq** [comment]
How difficult can it be to solve this issue so that all the interested people could use it? I tried my best to write this little tutorial to help people configure an additional wg tunnel. Now i am stuck at this point and i can't make it work, i tried almost anything.
I have been chasing you for 3 weeks trying to make it work. I guess my understanding of networks brings me only up to this point, the last step must be made by someone else, i just don't have the competence.

If you have time to make this last step would be great.

**Comment by highfreq** [comment]
If you ever try to make it work i'll add it up to the tutorial for other to not have to run in the darkness for weeks.

**Comment by highfreq** [comment]
Sorry for all the people maybe waiting for me finishing the tutorial but i couldn't make it work, i tried everything it came to mind. Rebuilt test server and client countless times but it just doesn't work.
For all the people needing an extra tunnel to reach OMR net i am sorry :(

**Comment by highfreq** [comment]
@Ysurac i guess i can't have anymore help to have this new wireguard tunnel functioning......

**Comment by Ysurac** [comment]
It's not directly related to OpenMPTCProuter and I never used this, so I can't make a tutorial. And for now I don't have the time to test this.
As I said you should make some trace with tcpdump to find where is the problem, and enable shorewall log, you can find how in shorewall doc (if I remember it's in policy file of shorewall config), without any of that I can't do anything.

**Comment by highfreq** [comment]
ok, thanks

**Comment by ioogithub** [comment]
@highfreq were you ever able to get wireguard working? It looks like you were close but didn't quite finish. @grapexy has written what looks like a high quality tutorial on setting up wirguard on the VPS here: https://github.com/Ysurac/openmptcprouter/issues/1536.

I have a question about PostUp and PostDown rules that I am hoping to get some clarification on but otherwise it looks like a complete solution. I wanted to link it here for you and anyone else in the future that would like to add a wireguard solution to their OMR setup. I saw several other users asking about this in my research of the issues.

If anyone has validated @grapexy's method please update us.

**Comment by highfreq** [comment]
Anyone, by any chance succeeded in configuring the last part of my configuration?

**Comment by highvolt-dev** [comment]
@highfreq thanks for creating this tutorial. I got everything working in both directions!

### OMR (LAN) Config

Static IPv4 Route for interface `omrvpn` with Target `10.255.230.0` and IPv4 Netmask of `255.255.255.0` and IPv4 Gateway of `10.255.255.1` (VPS gt-tun0 IP)

### Server (VPS) Config

Here's my VPS wireguard server config in `/etc/wireguard/wg1.conf`:

```ini
[Interface]
Address = 10.255.230.1/24
SaveConfig = true
ListenPort = 64999
PrivateKey = REDACTED

[Peer]
PublicKey = REDACTED
AllowedIPs = 10.255.230.2/32
```

#### Shorewall Firewall Config

`/etc/shorewall/rules`

```
ACCEPT          net             $FW             udp     64999
```

`/etc/shorewall/policy`

```
vpn             vpn             ACCEPT
```

### Wireguard Client

I actually copied and pasted the above wireguard client config over the wireguard VPN using RDP!

Note: my OMR LAN is on `192.168.12.0/24`

```ini
[Interface]
PrivateKey = REDACTED
Address = 10.255.230.2/24
DNS = 1.1.1.1
MTU = 1408

[Peer]
PublicKey = REDACTED
AllowedIPs = 10.255.230.0/24, 192.168.12.0/24, 10.255.255.0/24
Endpoint = REDACTED_VPS_ADDRESS:64999
PersistentKeepalive = 25
```

### Troubleshooting
I used tcpdump on both the OMR and VPS side listening for ICMP traffic.

OMR and the VPS normally ping each other anyway. When pinging the remote wireguard client (`10.255.230.2`) from my LAN, I set the ping length to `100` which created 108 byte packets. I could then filter in tcpdump with the `greater` keyword and not get distracted by the noise of OMR and the VPS pinging each other.

WHEN LAN CLIENT `192.168.12.x` requests to ping remote wireguard client `10.255.230.2`

OMR tcpdump on `tun0` (`10.255.255.2`) receives the following

```
13:50:14.977561 IP 10.255.255.2 > 10.255.230.2: ICMP echo request, id 1, seq 353, length 108

#                       OMR-tun0-iface TO remote wireguard client (ICMP REQUEST)
```

VPS tcpdump on all interfaces receives the following


1. Inbound ICMP echo request from OMR

```
13:50:15.003170 gt-tun0 In  IP 10.255.255.2 > 10.255.230.2: ICMP echo request, id 1, seq 353, length 108

#		   VPS-gt-tun0-iface (10.255.255.1) from src 10.255.255.2
```

1. Outbound ICMP echo request to wireguard interface

```
13:50:15.003194 wg1   Out IP 10.255.255.2 > 10.255.230.2: ICMP echo request, id 1, seq 353, length 108

                   VPS-wireguard-iface (10.255.230.1) from src 10.255.255.2
```

Confirming the necessary route exists and works on the VPS:

```
ip route get from 10.255.255.1 to 10.255.230.2
10.255.230.2 from 10.255.255.1 dev wg1 uid 0
```

Conclusion (and actual fix): `10.255.255.0/24` must be added to `AllowedIPs` section of wireguard client configuration in addition to your OMR lan IP (`192.168.12.0/24`) for me.

Thanks to both @Ysurac and @highfreq for the combined solution.
